drivers: usb: stm32: fix support of STM32U5 OTG_HS with embedded PHY

mniestroj opened the pull request:
Introduce new binding `st,stm32u5-otghs-phy` for OTG_HS PHY. This allows to
configure clock source and handle STM32U5 specific OTG_HS PHY behavior in
driver implementation in a more readable way.

Move OTG_HS PHY clock selection (previously `<&rcc STM32_SRC_HSI48
ICKLK_SEL(0)>`) from OTG_HS node to OTG_HS PHY node.

Rename `USBPHYC_SEL` -> `OTGHS_SEL` which matches the definition in the stm32u5
`CCIPR2` register (RM0456 Rev 5, Section 11.8.47).

Support enabling OTG_HS PHY clock, which is bit 15 (`OTGHSPHYEN`) in
`RCC_AHB2ENR1`. Change OTG_HS clock to be bit 14 (`OTGEN`).

Calculate in runtime OTG_HS PHY clock source frequency. Try to match that
to supported (16, 19.2, 20, 24, 26, 32 MHz) frequencies and select proper
option with `HAL_SYSCFG_SetOTGPHYReferenceClockSelection()` API (instead of
hardcoded 16 MHz selection).

This is alternative implementation (but uses parts of it) to: #80249. Most important differences:
 * OTG_HS PHY clock source frequency is calculated in runtime, so no need to specify that manually
   in devicetree.
   - better developer experience, due to less options to be selected and aligned with each other in
     devicetree
   - automatically checks in runtime if selected frequency is supported

Fixes: #79823

Tested with `nucleo_u5a5zj_q` board with `zephyr/samples/subsys/usb/cdc_acm` sample, in following
scenarios:

 * default (`HSE` as 16MHz clock source): no devicetree modifications
 * `PLL1_P` as 16MHz clock source:

 ```
&pll1 {
    div-p = <20>;
};

&otghs_phy {
    clocks = <&rcc STM32_CLOCK(AHB2, 15U)>,
             <&rcc STM32_SRC_PLL1_P OTGHS_SEL(1)>;
};
 ```

 * `PLL1_P` as 32MHz clock source:

 ```
&pll1 {
    div-p = <10>;
};

&otghs_phy {
    clocks = <&rcc STM32_CLOCK(AHB2, 15U)>,
             <&rcc STM32_SRC_PLL1_P OTGHS_SEL(1)>;
};
 ```

 * `PLL1_P` as 20MHz clock source:

```
&pll1 {
    div-p = <16>;
};

&otghs_phy {
    clocks = <&rcc STM32_CLOCK(AHB2, 15U)>,
             <&rcc STM32_SRC_PLL1_P OTGHS_SEL(1)>;
};
```

 * `PLL1_P/2` as 20MHz clock source:

```
&pll1 {
    div-p = <8>;
};

&otghs_phy {
    clocks = <&rcc STM32_CLOCK(AHB2, 15U)>,
             <&rcc (STM32_SRC_PLL1_P | STM32_CLOCK_DIV(2)) OTGHS_SEL(3)>;
};
```

Depends on (and contains): #82637

Commit:
boards: nucleo_u5a5zj_q: add USB HS 2.0 support

This board has the required clock crystal (X4) and jumper settings
present to enable the USB 2.0 HS support.

* Enable the HSE clock (16MHz)
* Flip the PLL1 configuration over to use the HSE clock, but still
  outputting 160MHz to sysclk/apbclk.
* Add the USB HS device tree node.
* Update the board documentation.

Signed-off-by: Adrian Chadd <adrian.chadd@meta.com>

mniestroj committed:
drivers: usb: stm32: fix support of STM32U5 OTG_HS with embedded PHY

Introduce new binding "st,stm32u5-otghs-phy" for OTG_HS PHY. This allows to
configure clock source and handle STM32U5 specific OTG_HS PHY behavior in
driver implementation in a more readable way.

Move OTG_HS PHY clock selection (previously <&rcc STM32_SRC_HSI48
ICKLK_SEL(0)>) from OTG_HS node to OTG_HS PHY node.

Rename USBPHYC_SEL -> OTGHS_SEL which matches the definition in the stm32u5
CCIPR2 register (RM0456 Rev 5, Section 11.8.47).

Support enabling OTG_HS PHY clock, which is bit 15 (OTGHSPHYEN) in
RCC_AHB2ENR1. Change OTG_HS clock to be bit 14 (OTGEN).

Calculate in runtime OTG_HS PHY clock source frequency. Try to match that
to supported (16, 19.2, 20, 24, 26, 32 MHz) frequencies and select proper
option with HAL_SYSCFG_SetOTGPHYReferenceClockSelection() API (instead of
hardcoded 16 MHz selection).

Co-authored-by: Adrian Chadd <adrian.chadd@meta.com>
Signed-off-by: Adrian Chadd <adrian.chadd@meta.com>
Signed-off-by: Marcin Niestroj <m.niestroj@emb.dev>

erwango commented on dts/bindings/phy/st,stm32u5-otghs-phy.yaml:
```diff
@@ -0,0 +1,36 @@
+                 <&rcc STM32_SRC_PLL1_P OTGHS_SEL(1)>;
+
+        /* PLL1_P_CK/2 */
+        clocks = <&rcc STM32_CLOCK(AHB2, 15U)>,
+                 <&rcc STM32_SRC_PLL1_P OTGHS_SEL(3)>;
```
I would actually prefer we introduce a new `STM32_SRC_PLL1_P_DIV_2` source to avoid breaking `clock_control_get_rate()` api behavior (with your proposal it may not return the right rate anymore).

To make it reusable, we can think of adding a trick on the u32 value such as :
```
                 <&rcc (STM32_SRC_PLL1_P | DIV_2) OTGHS_SEL(3)>;
```
With the DIV_2 being extracted in `clock_control_get_rate()` and included in rate computation before returning (to make it functional with similar HSE/2, HSI48/2, HSI/256, MSIK/1024 and MSIK/4)

mniestroj replied:
Great feedback!

I've implemented something along those lines, which allows to use:

```
<&rcc (STM32_SRC_PLL1_P | STM32_CLOCK_DIV(2)) OTGHS_SEL(3)>;
```

erwango reviewed (changes requested):
Initial feedback:

I'm all in for the main idea of instantiating a distinct phy clock, thanks. I'm fine with the dynamic computation as well.
Though, one comment on the clocks binding "abuse".

mniestroj committed:
drivers: clock: stm32: support STM32_CLOCK_DIV()

Support specifying divided clock buses by introduction of
STM32_CLOCK_DIV(div) macro. This macro can be used in devicetree to define
clock source of peripherals.

HSE is selected in devicetree using:

   <&rcc STM32_SRC_HSE ...>;

HSE/2 can now be selected with:

   <&rcc (STM32_SRC_HSE | STM32_CLOCK_DIV(2)) ...>;

This allows to use clock_control_get_rate() API in peripherals in order to
get desired clock rate.

Signed-off-by: Marcin Niestroj <m.niestroj@emb.dev>

FRASTM commented:
Fixing the build error  on west build -p auto -b nucleo_u5a5zj_q samples/boards/st/mco  with CI, requires 
```patch
diff --git a/include/zephyr/dt-bindings/clock/stm32u5_clock.h b/include/zephyr/dt-bindings/clock/stm32u5_clock.h
index 4b9b563121e..be159575518 100644
--- a/include/zephyr/dt-bindings/clock/stm32u5_clock.h
+++ b/include/zephyr/dt-bindings/clock/stm32u5_clock.h
@@ -139,4 +139,11 @@
 #define MCO1_SEL(val)           STM32_MCO_CFGR(val, 0xF, 24, CFGR1_REG)
 #define MCO1_PRE(val)           STM32_MCO_CFGR(val, 0x7, 28, CFGR1_REG)
 
+/* MCO prescaler : division factor */
+#define MCO_PRE_DIV_1 0
+#define MCO_PRE_DIV_2 4
+#define MCO_PRE_DIV_3 5
+#define MCO_PRE_DIV_4 6
+#define MCO_PRE_DIV_5 7
+
 #endif /* ZEPHYR_INCLUDE_DT_BINDINGS_CLOCK_STM32U5_CLOCK_H_ */
```

erwango commented on drivers/usb/device/usb_dc_stm32.c:
```diff
@@ -333,7 +422,7 @@ static int usb_dc_stm32_clock_disable(void)
 		return -EIO;
 	}
-#if DT_HAS_COMPAT_STATUS_OKAY(st_stm32_otghs) && defined(CONFIG_SOC_SERIES_STM32U5X)
+#if DT_HAS_COMPAT_STATUS_OKAY(st_stm32u5_otghs_phy)
 	LL_AHB2_GRP1_DisableClock(LL_AHB2_GRP1_PERIPH_USBPHY);
 #endif
```
Isn't this a `clock_control_off(clk, (clock_control_subsys_t)&phy_pclken[0]);` now?

mniestroj replied:
Yes, it seems so. Should I replace it with `clock_control_off()` API?

erwango replied:
it would be nice.

mniestroj replied:
Done!

erwango reviewed (dismissed):
Thanks @mniestroj for the quick and complete update.

Only required change would be to update rcc bindings description with the new STM32_CLOCK_DIV() trick ?

Otherwiwse looks globaly good to me.

marwaiehm-st reviewed (approved):
LGTM

erwango commented:
I'd like @erikarn feedback, but  otherwise LGTM.

erwango commented:
One more question, maybe for @marwaiehm-st: 
How does it fit with usb next support ? 
At the minimum, we should be able to bring this back (or forward) to usb next driver. Does it sound doable ?
(Ideally, we should not break STM32U5 OTG_HS support on USB next, but it is possible it doesn't work well so far anyway)

marwaiehm-st commented:
@erwango 
Yes, it is feasible. I do not foresee a significant issues with integrating the new st,stm32u5-otghs-phy into the USB next driver.

erikarn commented:
lgtm, thanks for digging into the weeds on cleaning this stuff up!

erikarn commented:
(I'm sad I won't get a commit into zephyr this time, but I'm sure there'll be others to do. I'll give this a run myself when it hits the mainline tree and provide some further testing feedback!)

mniestroj commented:
> (I'm sad I won't get a commit into zephyr this time, but I'm sure there'll be others to do. I'll give this a run myself when it hits the mainline tree and provide some further testing feedback!)

One commit in this PR is yours, while I have acknowledged your input (with Co-authored-by) in the second one. This PR would not exist in this shape without your work #80249. So thanks a lot for that!

The reason I created this PR is because I spent some time testing and already got ideas on how to improve that and ask community for feedback. Overall I just wanted to push this forward as far as I could, as otherwise I would might not find time for it in the future :)

So thanks a lot @erikarn for your contribution!

erikarn commented:
> > (I'm sad I won't get a commit into zephyr this time, but I'm sure there'll be others to do. I'll give this a run myself when it hits the mainline tree and provide some further testing feedback!)
> 
> One commit in this PR is yours, while I have acknowledged your input (with Co-authored-by) in the second one. This PR would not exist in this shape without your work #80249. So thanks a lot for that!
> 
> The reason I created this PR is because I spent some time testing and already got ideas on how to improve that and ask community for feedback. Overall I just wanted to push this forward as far as I could, as otherwise I would might not find time for it in the future :)
> 
> So thanks a lot @erikarn for your contribution!

ship itttttttt i have a bunch more usb stuff on my needing to do list :-) :-)

jfischer-no commented on drivers/usb/device/usb_dc_stm32.c:
```diff
@@ -208,16 +208,67 @@ void HAL_PCD_SOFCallback(PCD_HandleTypeDef *hpcd)
+
+static const struct stm32_pclken phy_pclken[] = STM32_DT_CLOCKS(DT_INST_PHANDLE(0, phys));
+
+static int usb_dc_stm32u5_phy_clock_select(const struct device *const clk)
```
@erwango Why is it not in the clock control driver, and why do we have to deal with all this unrelated driver code here? How will the changes be synchronized with the new driver?

erwango replied:
I understand your concern, but this configuration is purely out of clock control handling and depends on USB.
Note that even isn't a "RCC" block configuration, but a "SYSCFG" (System configuration controller) one.
Clock controller driver doesn't have information about the USB context, it "just" enables/disables clocks and has the knowledge about the related rate. Configuring which clock should be used it out if its scope and depend on consuming IP.

> How will the changes be synchronized with the new driver?
For sure, we're planning to port this to the new driver (see my above comment), but I didn't want to make this blocking for @mniestroj.

erwango replied:
@jfischer-no Please comment.

erwango replied:
FYI, just created https://github.com/zephyrproject-rtos/zephyr/issues/82745 for tracking

erwango replied:
@jfischer-no PTAL. Your legitimate concern about backport is I think addressed. For the other point about clock-control architecture, while I disagree, can be discussed later on but let's avoid scope creep in current PR.

erwango replied:
> We also have drivers/usb/common, you can put code there that is used by the both drivers

Interesting thanks. We'll surely consider it. Adding to comment to `drivers/usb/common`

jfischer-no reviewed (changes requested):
In the last commit, remove `(Issue https://github.com/zephyrproject-rtos/zephyr/issues/79823)` from the commit title.
